omegaconf/dictconfig.py

```diff
@@ -220,11 +231,37 @@ def _validate_merge(self, value: Any) -> None:
         src_obj_type = OmegaConf.get_type(src)
 
         if dest._is_missing() and src._metadata.object_type is not None:
-            self._validate_set(key=None, value=_get_value(src))
+            self._validate_set(
+                key=None, value=_get_value(src), allow_assigning_superclass=True
+            )
 
         if src._is_missing():
             return
 
+        if is_structured_config(src_obj_type):
+            assert src_obj_type is not None
+            dest_ref_type = dest._metadata.ref_type
+            dest_target_type = (
+                dest_obj_type
+                if is_structured_config(dest_obj_type)
+                else dest_ref_type
+                if is_structured_config(dest_ref_type)
+                else None
+            )
+            if dest_target_type is not None:
+                validation_warning = (
+                    issubclass(dest_target_type, src_obj_type)
+                    and dest_target_type is not src_obj_type
+                )
+                if validation_warning:
+                    msg = (
+                        f"Merging type {type_str(src_obj_type)} into subclass type"
+                        f" {type_str(dest_target_type)} is discouraged. See"
+                        " https://github.com/omry/omegaconf/issues/TODO"
```

Comment:
I'll create an issue with some details...

Reply:
I'm going to delay creating the issue until it's clear whether we'll need to merge this PR.